pkg/x509.go

```diff
@@ -210,40 +225,84 @@ func (client *Crypto) RenewTLSCertificate(entity types.LegalEntity) (*x509.Certi
 	entity := vendorEntity()
-	log.Logger().Debugf("Generating '%s' certificate for entity: %s", OAuthCertificateQualifier, entity)
+	log.Logger().Debugf("Generating '%s' certificate for entity: %s", qualifier, entity)
+
+	if err := validatePublicKeyRequirements(publicKey); err != nil {
+		return nil, err
+	}
 
 	caKey := types.KeyForEntity(entity)
 	caCertificate, err := client.Storage.GetCertificate(caKey)
 	if err != nil || caCertificate == nil {
-		return nil, nil, fmt.Errorf("unable to retrieve CA certificate %s: %v", caKey, err)
+		return nil, fmt.Errorf("unable to retrieve CA certificate %s: %v", caKey, err)
 	}
 	if len(caCertificate.Subject.Organization) == 0 {
-		return nil, nil, fmt.Errorf("subject of CA certificate %s doesn't contain 'O' component", caKey)
+		return nil, fmt.Errorf("subject of CA certificate %s doesn't contain 'O' component", caKey)
 	}
 	if len(caCertificate.Subject.Country) == 0 {
-		return nil, nil, fmt.Errorf("subject of CA certificate %s doesn't contain 'C' component", caKey)
+		return nil, fmt.Errorf("subject of CA certificate %s doesn't contain 'C' component", caKey)
 	}
 
 	var certificate *x509.Certificate
-	var privateKey *ecdsa.PrivateKey
-	if privateKey, err = generateECKeyPair(); err != nil {
-		return nil, nil, errors2.Wrapf(err, "unable to generate key pair for new %s certificate", OAuthCertificateQualifier)
-	}
-	csr, err := cert.CSRFromVendorCA(caCertificate, CACertificateQualifier, OAuthCertificateQualifier, &privateKey.PublicKey)
+
+	csr, err := cert.CSRFromVendorCA(caCertificate, CACertificateQualifier, qualifier, publicKey)
 	if err != nil {
-		return nil, nil, fmt.Errorf("unable to construct CSR: %w", err)
+		return nil, fmt.Errorf("unable to construct CSR: %w", err)
 	}
 
 	// in this case we don't serialize the CSR so the extra extensions need to be copied to the extensions
 	csr.Extensions = csr.ExtraExtensions
 
-	certificateAsBytes, err := client.signCertificate(csr, caKey, RFC003CertificateProfile, false)
+	certificateAsBytes, err := client.signCertificate(csr, caKey, profile, false)
 	if err != nil {
-		return nil, nil, errors2.Wrapf(err, "unable to generate %s certificate %s", OAuthCertificateQualifier, caKey)
+		return nil, errors2.Wrapf(err, "unable to generate %s certificate %s", qualifier, caKey)
 	}
 	certificate, err = x509.ParseCertificate(certificateAsBytes)
+	if err != nil {
+		return nil, err
+	}
+	return certificate, nil
+}
+
+func (client *Crypto) generateVendorTLSCertificate(publicKey crypto.PublicKey) (*x509.Certificate, error) {
+	return client.generateVendorCertificate(publicKey, TLSCertificateQualifier, RFC003TLSCertificateProfile)
+}
+
+func (client *Crypto) generateVendorEphemeralSigningCertificate() (*x509.Certificate, crypto.Signer, error) {
+	var privateKey *ecdsa.PrivateKey
+	var err error
+	if privateKey, err = generateECKeyPair(); err != nil {
+		return nil, nil, errors2.Wrapf(err, "unable to generate key pair for new %s certificate", OAuthCertificateQualifier)
```

Comment:
is this qualifier still right since you replaced it with `qualifier` above?

Reply:
yes since the entire func is just for signing purposes